File: Lucy/2025/2-W03/206_Reverse_Linked_List.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

// 두 번째 풀이(모범답안 참고)
function reverseList(head: ListNode | null): ListNode | null {
  let current = head;
  let prev = null;

  while (current) {
    const next = current.next;
    current.next = prev;
    prev = current;
    current = next;
  }

  return prev;
}

// 첫 번째 풀이
function reverseList(head: ListNode | null): ListNode | null {
  if (!head) return null;

  const nodes = [];
  let currentNode = head;

  while (currentNode) {
    nodes.push(currentNode.val);
    currentNode = currentNode.next;
  }

  nodes.reverse();

  const reversed = new ListNode(nodes[0]);
  currentNode = reversed;

  for (let i = 1; i < nodes.length; i++) {
    currentNode.next = new ListNode(nodes[i]);
    currentNode = currentNode.next;
  }

  return reversed;
}

File: Gombang/[LeetCode75] - [Linked List]/206_Reverse_Linked_List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {

        if (head == nullptr || head->next == nullptr)
            return head;

        ListNode* temp = head;
        stack<ListNode*> nodeStack; 
        // 스택에 하나씩 담는다.
        while (temp != nullptr)
        {
            nodeStack.push(temp);
            temp = temp->next;
        }

        ListNode* newHead = nodeStack.top();
        nodeStack.pop();

        temp = newHead;
        // 반대로 스택에서 꺼내면서 링크를 연결해준다.
        while (nodeStack.empty() == false)
        {
            temp->next = nodeStack.top();
            nodeStack.pop();
            temp = temp->next;
        }

        temp->next = nullptr;

        return newHead;
    }
};